### PR TITLE
Split options and flags in kubectl config sub-commands

### PR DESCRIPTION
**Pull request by mpuckett159**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Part of ongoing effort to split options and flags out in kubectl commands.

#### Special notes for your reviewer:

The way I have put this together we have three main categories of functions:

Commands that takes flags
- New\<subcommand\>Flags - This function creates a new initialized set of flags passing in the config file access and io streams objects at a minimum. All other flags are granted default values here.
- AddFlags - This function adds the flags to the `cobra.Command` object.
- ToOptions - This function takes the arguments provided by the user, the flags provided by the user, the default values of the flags, and composes all of this into a set of subcommand options that are returned which will then be later used by the run command.

Commands that do not take flags
- New\<subcommand\>Options - This function creates a new initialized set of options, passing the config file access and io streams objects at a minimum.
- Complete - This function takes the arguments provided by the user that would normally be added to the options object in the ToOptions function and adds them to the options object.

Utility functions
- Validate - Sometimes it is necessary to apply complicated logic to verify that an options object is both complete and valid based on user inputs. This function can accomplish this step if necessary.

#### Does this PR introduce a user-facing change?
No

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
See following issue for more details.

https://github.com/kubernetes/kubectl/issues/1046

### CODE REVIEW

**Comment by mpuckett159** [comment]
Alright as of this comment the subcommands in the config subcommand that use flags have been reworked to conform to the split flags/options model laid out in the wait subcommand. The list of changed files in the PR will tell you which that is.

All config subcommand tests pass, however, the config_test.go tests do not due to the way the tests are being run, not due to any underlying instability in the code itself. I'm not 100% sure but I think I need to rewrite the tests completely for it because of the method being used to ignore returned errors, which effectively allows the command to continue on error. This results in a null pointer error due to the new flags to options process.

Anyway, just fyi if people are reviewing before I get that fixed up.

**Comment by mpuckett159** [comment]
Ok turns out not all tests pass sans `config_test.go` tests, easy fix though.

**Comment by mpuckett159** [comment]
I believe integration is just a flake, but not going to retest until I get the `config_test.go` junk cleaned up.

**Comment by mpuckett159** [comment]
Well that's strange all the unit tests pass for me locally. Back to the drawing board.

**Comment by mpuckett159** [comment]
Oh good my favorite, do nothing and the problem goes away.

So now that I have a fix for the commands that required rewritting some of their logic, I'm gonna clean up the rest, then clean up the commit history and tag this as ready for review.

**Comment by mpuckett159** [comment]
Still not fully done but mostly done, really need to get the set commands redone and standardized, and then dig into getting the config tests done a bit better.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

**Comment by mpuckett159** [comment]
I know these are probably automated messages from dims but for clarity this WIP is still (intending to be) worked on and completed. I have my notes on what needs to be updated and fixed and just need to finish the work.

**Comment by mpuckett159** [comment]
/release-note-none
/priority backlog
/triage accepted

I think I need to stop tweeking this an just explain the reasoning.

**Comment by mpuckett159** [comment]
This should almost certainly have been split into smaller PRs sorry to whoever ends up reviewing this 😬

File: staging/src/k8s.io/kubectl/pkg/cmd/config/config_test.go [hunk 1]
```diff
@@ -364,24 +371,24 @@ func TestTokenAndBasicDisallowed(t *testing.T) {
...
-			if code != 1 {
-				t.Errorf("The exit code is %d, expected 1", code)
-			}
-
-			expectedOutputs := []string{"--token", "--username"}
-			test.checkOutput(e, expectedOutputs, t)
-		})
-
+	// Allow for command errors in testing
+	if os.Getenv("CRASH") == "1" {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/config_test.go:375]: I did not know that environment variable, is it something widely used in k8s and what does it for?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/current_context.go [hunk 2]
```diff
@@ -42,35 +37,56 @@ var (
...
+	if len(args) != 0 {
+		return fmt.Errorf("unexpected args: %v", args)
+	}
+	return nil
+}
+
 // RunCurrentContext performs the execution of 'config current-context' sub command
-func RunCurrentContext(out io.Writer, options *CurrentContextOptions) error {
-	config, err := options.ConfigAccess.GetStartingConfig()
+func (o *CurrentContextOptions) RunCurrentContext() error {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/current_context.go:78]: Maybe now we can safely rename this function to `Run`, since we already know that this is `CurrentContextOptions` function?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/current_context.go [hunk 3]
```diff
@@ -42,35 +37,56 @@ var (
...
 	}
 
 	if config.CurrentContext == "" {
-		err = fmt.Errorf("current-context is not set")
-		return err
+		return fmt.Errorf("current-context is not set")
 	}
 
-	fmt.Fprintf(out, "%s\n", config.CurrentContext)
+	if _, err := fmt.Fprintf(o.IOStreams.Out, "%s\n", config.CurrentContext); err != nil {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/current_context.go:88]: How this function can get error?, even if it gets somehow an error, should it block the command execution?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/delete_cluster.go [hunk 4]
```diff
@@ -44,43 +53,50 @@ func NewCmdConfigDeleteCluster(out io.Writer, configAccess clientcmd.ConfigAcces
...
+		return fmt.Errorf("unexpected args: %v", args)
 	}
+	o.Cluster = args[0]
+	return nil
+}
 
-	configFile := configAccess.GetDefaultFilename()
-	if configAccess.IsExplicitFile() {
-		configFile = configAccess.GetExplicitFile()
+func (o *DeleteClusterOptions) RunDeleteCluster() error {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/delete_cluster.go:81]: Same above and all the others.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/delete_context.go [hunk 5]
```diff
@@ -44,47 +53,56 @@ func NewCmdConfigDeleteContext(out, errOut io.Writer, configAccess clientcmd.Con
...
+		IOStreams:    ioStreams,
 	}
+}
 
-	args := cmd.Flags().Args()
+func (o *DeleteContextOptions) Complete(args []string) error {
 	if len(args) != 1 {
-		cmd.Help()
-		return nil
+		return fmt.Errorf("unexpected args: %v", args)
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/delete_context.go:75]: Isn't this changing the behavior?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/delete_user.go [hunk 6]
```diff
@@ -38,27 +37,14 @@ var (
 
 // DeleteUserOptions holds the data needed to run the command
 type DeleteUserOptions struct {
-	user string
+	User string
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/delete_user.go:40]: What is the reason of exporting this field?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/delete_user.go [hunk 7]
```diff
@@ -67,56 +53,51 @@ func NewCmdConfigDeleteUser(streams genericclioptions.IOStreams, configAccess cl
...
 }
 
-// Validate ensures the command has enough info to run
-func (o *DeleteUserOptions) Validate() error {
-	_, ok := o.config.AuthInfos[o.user]
-	if !ok {
-		return fmt.Errorf("cannot delete user %s, not in %s", o.user, o.configFile)
+// RunDeleteUser performs the command
+func (o *DeleteUserOptions) RunDeleteUser() error {
+	config, configFile, err := loadConfig(o.ConfigAccess)
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/delete_user.go:84]: Is this refactoring related to the options and flags split?

**Review by ardaguclu — COMMENTED** [review]
Thanks for the PR, this is a good effort and generally all looks good.

I think, there are different kind of changes are all baked in this PR. Would it be better to open separate PRs for each of them and this PR only includes options->flag split?

Because, we need to ensure that unit tests are verifying this changes but in this PR unit tests are also being changed inevitably. Also would you mind adding more integrations tests to assure that everything works as expected.

**Comment by mpuckett159** [comment]
Yeah, you know, honestly, you're right. I let this one get away from me. I'm going to break this up into smaller PRs. This thing is a nightmare. I let the scope creep way too far.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/config_test.go (hunk 1 above)
Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/config_test.go:375]: This is a way to allow for commands to issue OS exit codes when they fail and keep testing them. Go doesn't really advise doing this, but it's possible, and I was struggling with the method used before for unit tests in these files, so I converted them to this method. What I intend to do is to just move these kinds of tests to the integration testing because that is generally what is advised these days for testing failure states like this. I'll need to do that in a separate PR as noted in my other comment.
